# WGS_anaylsis_workflow

The repository contains the analysis workflow used to analyze whole-genome sequencing datasets generated during this study: [add publication DOI here].

The workflow is based around several tools that have been incorporated into 3 Snakemake pipelines (Fig. 1).
The aim of the first pipeline (Snakefile_GATK) is to generated an alignment file ({sample}/{sample}_WGS_hg19_recalibrated.sorted.bam) and two normalized variant call files (VCFs), generated by HaplotypeCaller and DeepVariant, respectively.

The second pipeline (Snakefile_SV_pipeline) produces strctural variant calls using six different base tools (Lumpy, Manta, GRIDSS, WHAMg, CNVpytor, and Delly2), which are finally merged into a single VCF file (Fig. 2).

The third pipeline (Snakefile_Nirvana) integrates Nirvana to annotate all variants and finally merges and filters the annotated variants based on a genes list of interest (Fig. 1).
