# WGS_anaylsis_workflow

The repository contains the analysis workflow used to analyze whole-genome sequencing datasets generated during this study: [**<em>add publication DOI here</em>**].
The workflow is based around several tools that have been incorporated into 3 Snakemake pipelines (Fig. 1).


<img src="/Figure_1_github.png" alt="Fig. 1: WGS analysis workflow overview" width="800"/>

The aim of the first pipeline (**Snakefile_GATK**) is to generated an alignment file (<code>{sample}/{sample}_WGS_hg19_recalibrated.sorted.bam</code>) and two normalized variant call files (VCFs), generated by <code>[HaplotypeCaller](https://gatk.broadinstitute.org/hc/en-us/articles/21905025322523-HaplotypeCaller)</code> and <code>[DeepVariant](https://github.com/google/deepvariant)</code>, respectively.

The second pipeline (**Snakefile_SV_pipeline**) produces strctural variant calls using six different base tools (<code>[lumpy](https://github.com/arq5x/lumpy-sv)</code>, <code>[Manta](https://github.com/Illumina/manta)</code>, <code>[Delly2](https://github.com/dellytools/delly)</code>, <code>[CNVpytor](https://github.com/abyzovlab/CNVpytor)</code>, <code>[GRIDSS](https://github.com/PapenfussLab/gridss)</code>, and <code>[WHAM](https://github.com/zeeev/wham)</code>), which are finally merged into a single VCF file (Fig. 2).


<img src="/Figure_2_github.png" alt="Fig. 2: Structural variants discovery workflow" width="800"/>


The third pipeline (**Snakefile_Nirvana**) integrates Nirvana to annotate all variants and finally merges and filters the annotated variants based on a genes list of interest (Fig. 1).
